### PR TITLE
Add check codeowners on CI

### PR DESCRIPTION
**Pull request by eundoosong**
Signed-off-by: Eundoo Song <eundoo.song@gmail.com>

**Description:** <Describe what has changed.>

- To check every component has a CODEOWNERS line, `check_code_owner_existence` is added.
  - To get the list of every component, I used `find` to search for `go.mod` as commented in https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/3871#issuecomment-869700055. 
  But there are more components not included in CODEOWNERS in addition to #3870 as you see below result.  Are these all required to be in CODEOWNERS?
- To check every component listed on CODEOWNERS actually exists (to prevent typos), `check_component_existence` is added.
- To call these two on CI,  a separate job `check-codeowners` is added in build-and-test.yml and called after `setup-environment`

#### The result of `check_code_owner_existence`
```
"exporter/awscloudwatchlogsexporter" not included in CODEOWNERS
"exporter/influxdbexporter" not included in CODEOWNERS
"exporter/observiqexporter" not included in CODEOWNERS
"extension/fluentbitextension" not included in CODEOWNERS
"extension/oauth2clientauthextension" not included in CODEOWNERS
"extension/observer/ecsobserver" not included in CODEOWNERS
"extension/observer/hostobserver" not included in CODEOWNERS
"extension/observer/k8sobserver" not included in CODEOWNERS
"extension/storage" not included in CODEOWNERS
"internal/aws/awsutil" not included in CODEOWNERS
"internal/aws/containerinsight" not included in CODEOWNERS
"internal/aws/k8s" not included in CODEOWNERS
"internal/aws/metrics" not included in CODEOWNERS
"internal/aws/xray" not included in CODEOWNERS
"internal/aws/xray/testdata/sampleapp" not included in CODEOWNERS
"internal/aws/xray/testdata/sampleserver" not included in CODEOWNERS
"internal/common" not included in CODEOWNERS
"internal/kubelet" not included in CODEOWNERS
"internal/tools" not included in CODEOWNERS
"pkg/batchperresourceattr" not included in CODEOWNERS
"pkg/batchpersignal" not included in CODEOWNERS
"pkg/experimentalmetricmetadata" not included in CODEOWNERS
"processor/cumulativetodeltaprocessor" not included in CODEOWNERS
"processor/metricsgenerationprocessor" not included in CODEOWNERS
"receiver/fluentforwardreceiver" not included in CODEOWNERS
"receiver/googlecloudpubsubreceiver" not included in CODEOWNERS
"receiver/influxdbreceiver" not included in CODEOWNERS
"receiver/kafkametricsreceiver" not included in CODEOWNERS
"receiver/memcachedreceiver" not included in CODEOWNERS
"receiver/nginxreceiver" not included in CODEOWNERS
"receiver/simpleprometheusreceiver/examples/federation/prom-counter" not included in CODEOWNERS
"receiver/syslogreceiver" not included in CODEOWNERS
"receiver/tcplogreceiver" not included in CODEOWNERS
"receiver/udplogreceiver" not included in CODEOWNERS
"receiver/zookeeperreceiver" not included in CODEOWNERS
"testbed" not included in CODEOWNERS
"testbed/mockdatareceivers/mockawsxrayreceiver" not included in CODEOWNERS
there are 37 components not included in CODEOWNERS
```

#### The result of `check_component_existence`
```
"receiver/googlecloudpubsubexporter/" does not exist as specified in CODEOWNERS
there are 1 component(s) that do not exist as specified in CODEOWNERS
```

**Link to tracking Issue:** <Issue number if applicable>
#3871 

**Testing:** <Describe what testing was performed and which tests were added.>

### CODE REVIEW

File: .github/workflows/scripts/check-codeowners.sh [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+}
+
+check_component_existence() {
+  NOT_EXIST_COMPONENTS=0
+  while IFS= read -r line
+  do
+    if [[ $line =~ ^[^#\*] ]]; then
+      COMPONENT_PATH=$(echo "$line" | cut -d" " -f1)
+      if [ ! -d "$COMPONENT_PATH" ]; then
+        echo "\"$COMPONENT_PATH\" not existed as specified in CODEOWNERS"
```
Comment (Aneurysm9) [inline .github/workflows/scripts/check-codeowners.sh:26]:
```suggestion
        echo "\"$COMPONENT_PATH\" does not exist as specified in CODEOWNERS"
```

File: .github/workflows/scripts/check-codeowners.sh [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+        ((NOT_EXIST_COMPONENTS=NOT_EXIST_COMPONENTS+1))
+      fi
+    fi
+  done <"$CODEOWNERS"
+  echo "there are $NOT_EXIST_COMPONENTS components not existed as specified in CODEOWNERS"
```
Comment (Aneurysm9) [inline .github/workflows/scripts/check-codeowners.sh:31]:
```suggestion
  echo "there are $NOT_EXIST_COMPONENTS component(s) that do not exist as specified in CODEOWNERS"
```

**Comment by bogdandrutu** [comment]
@Aneurysm9 please let me know when this is ready to merge

**Review by mx-psi — CHANGES_REQUESTED** [review]
Thanks again for working on this @eundoosong :) I will update my issue with some components that are missing from the list of orphan components, there were indeed quite a few missing.

I think there is one case that I missed on my issue description and needs to be taken into account here: sometimes several Go modules are listed in a single entry on CODEOWNERS by referring to a parent folder. For example, in your script output we can see

```
"internal/aws/awsutil" not included in CODEOWNERS
"internal/aws/containerinsight" not included in CODEOWNERS
"internal/aws/k8s" not included in CODEOWNERS
"internal/aws/metrics" not included in CODEOWNERS
"internal/aws/xray" not included in CODEOWNERS
"internal/aws/xray/testdata/sampleapp" not included in CODEOWNERS
"internal/aws/xray/testdata/sampleserver" not included in CODEOWNERS
```

But these should not show up since `internal/aws` is listed on the CODEOWNERS file: https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/e25ce99cd24cfcf0433ee39c1d62f4bf3be74396/.github/CODEOWNERS#L50
The same thing happens with e.g. `extension/observer/`.

That means your script should check if there is a path on CODEOWNERS which is a _prefix_ of the Go module path instead of just checking the full path.

**Comment by eundoosong** [comment]
> Thanks again for working on this @eundoosong :) I will update my issue with some components that are missing from the list of orphan components, there were indeed quite a few missing.
> 
> I think there is one case that I missed on my issue description and needs to be taken into account here: sometimes several Go modules are listed in a single entry on CODEOWNERS by referring to a parent folder. For example, in your script output we can see
> 
> ```
> "internal/aws/awsutil" not included in CODEOWNERS
> "internal/aws/containerinsight" not included in CODEOWNERS
> "internal/aws/k8s" not included in CODEOWNERS
> "internal/aws/metrics" not included in CODEOWNERS
> "internal/aws/xray" not included in CODEOWNERS
> "internal/aws/xray/testdata/sampleapp" not included in CODEOWNERS
> "internal/aws/xray/testdata/sampleserver" not included in CODEOWNERS
> ```
> 
> But these should not show up since `internal/aws` is listed on the CODEOWNERS file:
> 
> https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/e25ce99cd24cfcf0433ee39c1d62f4bf3be74396/.github/CODEOWNERS#L50
> 
> 
> The same thing happens with e.g. `extension/observer/`.
> That means your script should check if there is a path on CODEOWNERS which is a _prefix_ of the Go module path instead of just checking the full path.

Ok thanks for your review and I'll look into how your requirement works in the script.

**Comment by mx-psi** [comment]
@eundoosong do you plan on working further on this in the future?

**Comment by hickeyma** [comment]
Superseded by #9751

**Pull request closed without merging**